docs(CONTRIBUTING): revert should be a valid type

stevemao opened the pull request:
EG: https://github.com/angular/angular.js/commit/462f444b06ae5cad3ccb761b1dba7131df01a655

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/angular/angular.js/12032)

<!-- Reviewable:end -->

petebacondarwin commented:
Well, strictly, `revert` is a modifier that you stick on the front of another commit's message.

petebacondarwin commented:
Can you describe it like that instead?

stevemao commented:
Since there is only one "modifier", should I just say `revert`?
I'm not 100% sure how you wanna write it but I'll push up something.

stevemao commented:
Also I'm maintaining the module [conventional-changelog](https://github.com/ajoslin/conventional-changelog) used by you guys.
what do you expect conventional-changelog to do to this? I would simply parse it as a `type` so it will look the same as other types. 

```md
## Bug Fixes
...

## Reverts
...
```

petebacondarwin commented:
Well at the moment, we are not actually using your package, Instead we use an script inside the angular.js project, `changelog.js`. And in this script you can see that we ignore all commits that do not start with `fix`, `feat`, `perf` or contain the word `BREAKING`. See https://github.com/angular/angular.js/blob/master/changelog.js#L194

The general idea, though, as far I am concerned, is that reverted commits should be removed from the changelog. So assuming that the original commit is also in the current release then we should simply ignore both commits. If it is reverting from a previous release, then I agree that we should have some kind of `## Reverts` section.

stevemao commented:
Yup, that script is actually the very first version of conventional-changelog. You guys should definitely use the package once I release `v0.1.x`. It's much better :)
I agree with you. So far it's not smart enough to remove both commits within one release. Thanks for the feedback @petebacondarwin 

Let me know what you think of the docs.

petebacondarwin commented on CONTRIBUTING.md:
````diff
@@ -192,7 +192,7 @@ Each commit message consists of a **header**, a **body** and a **footer**.  The
 
 ```
-<type>(<scope>): <subject>
+(revert: )<type>(<scope>): <subject>
````
Hmm, the parentheses here on `(revert: )` are because it is optional, right? But the parentheses on `(<scope>)` are because they are needed in the string. Perhaps we should just not include the `revert` bit in this and just explain about reverts as you do very clearly below.

stevemao replied:
Oh right. I was thinking the scope is actually optional. Maybe we should note what's optional what's not?

petebacondarwin commented on CONTRIBUTING.md:
```diff
@@ -202,6 +202,10 @@ format that includes a **type**, a **scope** and a **subject**:
+### Revert
+
+If the commit reverts a previous commit, it should begin with `revert: `, followed by the header of the reverted commit. In the body it should say: `This reverts commit <hash>.`
+
```
(where the hash is the SHA of the commit being reverted) :-)

stevemao commented:
One thing I found about the commit message guide is that it's actually not very strict:
1. It doesn't say what's optional what's not.
2. The exact format of a part.
For example, it doesn't say how **Breaking Changes** should be written. The sentense should start with `Breaking Change: ` and there are optional two newlines between `Breaking Change: ` and the content of it.

stevemao commented:
This applies to the google doc you link to in the file. It gives more examples but still not very strict.

stevemao committed:
docs(CONTRIBUTING): how to write a breaking change

stevemao committed:
docs(CONTRIBUTING): state what is mandatory or optional

stevemao commented:
I added two commits just to support my points (Maybe they should be in different PR but they kinda related and I need your opinions). Let me know what you think. Thanks.

stevemao committed:
docs(CONTRIBUTING): revert is a modifier

EG: https://github.com/angular/angular.js/commit/462f444b06ae5cad3ccb761b1dba7131df01a655

stevemao commented:
ping @petebacondarwin

gkalpak commented:
Fun fact: According to the `validate-commit-msg.js` file, `revert` is considered a valid type since Nov 2012 (and there are tests that verify it - which probably no one runs :smiley:).

stevemao commented:
hm... lol. `conventional-changelog` treats `revert` as a type too (if it reverts a commit that's not in the current release) and it works well. There is no need to make an exceptional rule for it.

for `validate-commit-msg.js` I think it should verify the reverted commit actually exists.

petebacondarwin commented:
LGTM!

stevemao commented:
Thanks @petebacondarwin , @gkalpak anything you would like to add here?

stevemao commented:
Would be great if you guys could update the docs on google docs
https://docs.google.com/document/d/1QrDFcIiPjSLDn3EL15IJygNPiHORgU1_OOAqWjiDU5Y/edit#

Or, even better, create a git repo just for the conventions?

stevemao commented:
Thanks @petebacondarwin

petebacondarwin commented:
@stevemao I have also updated the google doc.

stevemao commented:
Nice, will check it out
